src/codec/syslog.rs

```diff
@@ -0,0 +1,479 @@
+
+        if let Some(o) = data.get("structured_data") {
+            if let Some(sd) = o.as_object() {
+                let mut elem = String::new();
+                for (id, params) in sd.iter() {
+                    elem.push('[');
+                    elem.push_str(&id.to_string());
+                    if let Some(v) = params.as_array() {
+                        for key_value in v.iter() {
+                            if let Some(o) = key_value.as_object() {
+                                for (k, v) in o {
+                                    if let Some(v) = v.as_str() {
+                                        elem.push(' ');
+                                        elem.push_str(&k.to_string());
+                                        elem.push('=');
+                                        elem.push('"');
+                                        elem.push_str(v);
+                                        elem.push('"');
+                                    } else {
+                                        return Err(ErrorKind::InvalidSyslogData(
+                                            "Invalid structured data: param value not a string",
+                                        )
+                                        .into());
+                                    }
+                                }
+                            } else {
+                                return Err(ErrorKind::InvalidSyslogData(
+                                    "Invalid structured data: param's key value pair not an object",
+                                )
+                                .into());
+                            }
+                        }
+                    } else {
+                        return Err(ErrorKind::InvalidSyslogData(
+                            "Invalid structured data: params not an array of objects",
+                        )
+                        .into());
+                    }
+                    elem.push(']');
+                }
+                result.push(elem);
+            } else {
+                return Err(ErrorKind::InvalidSyslogData(
+                    "Invalid structured data: structured data not an object",
+                )
+                .into());
+            }
+        } else if let Protocol::RFC5424(_) = protocol {
+            result.push(String::from("-"));
+        }
+
+        if let Some(msg) = data.get_str("msg") {
+            result.push(msg.to_owned());
+        }
+
+        Ok(result.join(" ").as_bytes().to_vec())
+    }
+
+    #[cfg(not(tarpaulin_include))]
+    fn boxed_clone(&self) -> Box<dyn Codec> {
+        Box::new(self.clone())
+    }
+}
+
+// Function used to resolve the year for syslog messages that don't include the year.
+// If the current month is January, and the syslog message is for December, it will take the previous year.
+// Otherwise, take the current year.
+fn resolve_year((month, _date, _hour, _min, _sec): IncompleteDate) -> i32 {
```

Comment:
wouldn't it make sense to return the previous year for all month that are larger then the current month?